File: scala-concurrency/src/main/scala/net/ansalon/examples/concurrency/ch06/EnergyConsumer.scala
package net.ansalon.examples.concurrency.ch06

import scala.actors.Actor._

object EnergyConsumer {
  val source = EnergySource.create

  def main(args: Array[String]): Unit = {
    println("Energy level at start: " + source.unitsAvailable)

    val caller = self
    for (i <- 1 to 10) actor {
      for (j <- 1 to 7) source.consume(1)
      caller ! true
    }

    for (i <- 1 to 10) { receiveWithin(1000) { case message => } }
    
    println("Energy level at end: " + source.unitsAvailable)
    println("Usage: " + source.usages)
    
    source.stop
  }
}

File: programming-concurrency/src/main/java/net/ansalon/examples/concurrency/ch06/EnergyConsumer.java
package net.ansalon.examples.concurrency.ch06;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class EnergyConsumer {
	private static final AkkaEnergySource energySource = AkkaEnergySource.create();
	
	public static void main(String[] args) {
		System.out.println("Energy level at start: "
				+ energySource.getUnitsAvailable());
		List<Callable<Object>> tasks = new ArrayList<Callable<Object>>();
		for (int i = 0; i < 10; i++) {
			tasks.add(new Callable<Object>() {
				public Object call() {
					for (int j = 0; j < 7; j++)
						energySource.consume(1);
					return null;
				}
			});
		}
		
		final ExecutorService service = Executors.newFixedThreadPool(10);
		try {			
			service.invokeAll(tasks);
			System.out.println("Energy level at end: " + energySource.getUnitsAvailable());
			System.out.println("Usage: " + energySource.getUsageCount());			
		} catch (InterruptedException e) {			
			e.printStackTrace();
		} finally {
			energySource.stop();
			service.shutdown();
		}
	}
}
